src/Tempest/Http/src/Routing/Construction/RouteTreeNode.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Http\Routing\Construction;
+
+use Tempest\Http\Route;
+
+/**
+ * @internal
+ */
+final class RouteTreeNode
+{
+    /** @var array<string, RouteTreeNode> */
+    private array $staticPaths = [];
+
+    /** @var array<string, RouteTreeNode> */
+    private array $dynamicPaths = [];
+
+    private ?MarkedRoute $leaf = null;
+
+    private function __construct(
+        public readonly RouteTreeNodeType $type,
+        public readonly ?string $segment = null
+    ) {
+    }
+
+    public static function createRootRoute(): self
+    {
+        return new self(RouteTreeNodeType::Root);
+    }
+
+    public static function createParameterRoute(string $regex): self
+    {
+        return new self(RouteTreeNodeType::Parameter, $regex);
+    }
+
+    public static function createStaticRoute(string $name): self
+    {
+        return new self(RouteTreeNodeType::Static, $name);
+    }
+
+    public function addPath(array $pathSegments, MarkedRoute $markedRoute): void
+    {
+        if ($pathSegments === []) {
+            if ($this->leaf !== null) {
+                throw new DuplicateRouteException($markedRoute->route);
+            }
+
+            $this->leaf = $markedRoute;
+
+            return;
+        }
+
+        $currentPathSegment = array_shift($pathSegments);
+
+        $regexPathSegment = self::convertDynamicSegmentToRegex($currentPathSegment);
+
+        if ($currentPathSegment !== $regexPathSegment) {
+            $node = $this->dynamicPaths[$regexPathSegment] ??= self::createParameterRoute($regexPathSegment);
+        } else {
+            $node = $this->staticPaths[$regexPathSegment] ??= self::createStaticRoute($currentPathSegment);
+        }
+
+        $node->addPath($pathSegments, $markedRoute);
+    }
+
+    public static function convertDynamicSegmentToRegex(string $uriPart): string
+    {
+        $regex = '#\{'. Route::ROUTE_PARAM_NAME_REGEX . Route::ROUTE_PARAM_CUSTOM_REGEX .'\}#';
+
+        return preg_replace_callback(
+            $regex,
+            static fn ($matches) => trim($matches[2] ?? Route::DEFAULT_MATCHING_GROUP),
+            $uriPart,
+        );
+    }
+
+    private function regexSegment(): string
+    {
+        return match($this->type) {
+            RouteTreeNodeType::Root => '^',
+            RouteTreeNodeType::Static => "/{$this->segment}",
+            RouteTreeNodeType::Parameter => '/(' . $this->segment . ')',
+        };
+    }
+
+    public function toRegex(): string
```

Comment:
This uses recursion to construct the regex.

It will create nesting for all `$this->staticPaths` and `$this->dynamicPaths`. And if it is a leaf it will add an regex marker with the route `mark` to it.  We add the staticPaths before the dynamicPaths to give them preference.